specs/jsonschema-core.md

```diff
@@ -1012,21 +1013,22 @@ resolution until runtime, at which point it is resolved each time it is
 encountered while evaluating an instance.
 
 Together with `$dynamicAnchor`, `$dynamicRef` implements a cooperative extension
-mechanism that is primarily useful with recursive schemas (schemas that
-reference themselves). The extension point is defined with `$dynamicAnchor` and
-only exhibits runtime dynamic behavior when referenced with `$dynamicRef`.
+mechanism that is primarily useful to extend recursive schemas, where
+`$dynamicRef` defines the extension point, and `$dynamicAnchor` defines the
+target.
 
-The value of the `$dynamicRef` property MUST be a string which is a
-IRI reference that contains a valid [plain name fragment](#anchors). Resolved
-against the current IRI base, it indicates the schema resource used as the
-starting point for runtime resolution. This initial resolution is safe to
-perform on schema load.
+The value of the `$dynamicRef` property MUST be a valid
+[plain name fragment](#anchors).[^3]
 
-The schema to apply is the outermost schema resource in the [dynamic
-scope](#scopes) that defines a `$dynamicAnchor` that matches the plain name
-fragment in the initially resolved IRI.
+[^3]: `$dynamicAnchor` defines the anchor with plain text, e.g. `foo`; the
+`$dynamicRef` that references it uses a URI fragment syntax, e.g. `#foo`.
```

Comment:
> the footnote describes it as a fragment-only URI with plain name fragment

The footnote says it has that _syntax_, not that it _is_ an IRI fragment.

> It was part of my proposal that it should become just a plain name identifier because there's no need for it to be a URI without the initial resolution step

I assume you're talking about using the plain name in `$dynamicRef`?  If we don't use the IRI (I do have "URI" in the above text) fragment syntax in `$dynamicRef`, then it won't be backward compatible (if that matters to you here).

Reply:
> I assume you're talking about using the plain name in $dynamicRef?

Yes. Just the plain name not as part of a fragment-only IRI.

> If we don't use the IRI (I do have "URI" in the above text) fragment syntax in $dynamicRef, then it won't be backward compatible (if that matters to you here).

Removing the initial resolution step already makes it a breaking change, so I'm not too worried about introducing another. However, I recognize that there's a difference in the impact of the two changes. I doubt the initial resolution is used anywhere other than the "links" schema (used by the hyper-schema meta-schema), so there's almost no impact. The other change is minor (just remove the leading `#`), but affects every use of `$dynamicRef`.

Personally, I think that using IRI syntax for something that isn't an IRI is unnecessary and confusing and makes for a bad experience for people trying to learn this feature. I think it's worth fixing while we still can.

Reply:
Let's keep it as is for now.  If you can work up a voting issue, we can run it by the team.

Reply:
> Is there anything needed for this PR? I understand we have further changes coming, but this is a good step in the right direction.

I'm fine with leaving this value as a fragment-only IRI for now, but the text is still ambiguous at best about what the value of `$dynamicRef` should be. Right now it seems to say that it should be just a fragment identifier (not an IRI) and that contradicts the footnote that says it's an IRI. All other concerns I mentioned can be addressed later, but I think the spec should at least be self-consistent before we merge.

Reply:
What if we say `$dynamicRef` SHOULD be just the same name but MAY be formatted as a plain name fragment and that implementations MUST support both?

This kind of gets us into the same situation that we had when we started to forbid fragments in `$id`, but I think it's a good way to start the transition toward using the plain name.  I don't think this would be that great of a burden on implementations.

Reply:
Suggestion:

> The value of the `$dynamicRef` property MAY either be
> 
> - the exact string value as found in the corresponding `$dynamicAnchor`, or
> - the `$dynamicAnchor` value formatted as a valid [plain name fragment](#fragments).
> 
> Implementations MUST support both variations.[^3]
> 
> [^3]: Since `$dynamicRef`/`$dynamicAnchor` references are a closed system, distinct from `$ref`/`$id`/`$anchor` references, it is not necessary for `$dynamicRef` to use IRIs or their syntax. However, due to the large number of existing 2020-12 schemas which currently use an IRI fragment syntax in `$dynamicRef`, support for this formatting is still required. It is likely that this will eventually be phased out.

Reply:
Supporting two different syntaxes at the same time seems really confusing. Plus, making this change means that it will no longer be possible to use a dynamicRef to another document or even a different namespace in the same document, which seems like a huge step backwards.

Reply:
> Supporting two different syntaxes at the same time seems really confusing.

I agree, but I don't see another incremental way toward abandoning the IRI format for `$dynamicRef`, which seems to make sense to do.

> making this change means that it will no longer be possible to use a dynamicRef to another document or even a different namespace in the same document

I think you may have missed that `$dynamicRef` can no longer operate as `$ref`.  It can only target `$dynamicAnchor` moving forward.

Reply:
> What if we say `$dynamicRef` SHOULD be just the same name but MAY be formatted as a plain name fragment and that implementations MUST support both?

I think this is a great idea in normal circumstances, but considering where we are with the transition to the stable spec, I don't think this is the right approach. Once we do this release, we're committing to no breaking changes. We decided on a process for introducing breaking changes, but we should be doing everything we can to avoid having to use that process. Releasing the spec with a something that we already know needs to change is not ideal. I think we need to either change it and make it stable, or allow both and make it experimental until we're comfortable removing the deprecated syntax. I'm ok with either.

> * the $dynamicAnchor value formatted as a valid [plain name fragment](https://github.com/json-schema-org/json-schema-spec/pull/1555#fragments).

I don't think I've communicated well enough my concern about the term "plain name fragment" because you keep using that term the same way and I understand it to mean something different than you do. I understand this term to mean the fragment part of a URI that uses a plain name identifier. You clearly understand this to mean a fragment-only URI using a plain name identifier. This language needs to be more precise to avoid the ambiguity.

> However, due to the large number of existing 2020-12 schemas which currently use an IRI fragment syntax in `$dynamicRef`,

I don't think it's accurate to say there are a large number of schemas using `$dynamicRef`. We don't know for sure, but I think dynamic references are a very rarely used feature. I think only a small number of power users understand how to use it.

> making this change means that it will no longer be possible to use a dynamicRef to another document or even a different namespace in the same document

You can still express all of the same constraints you could before. Just the way you do it is a little different. The initial resolution step was only necessary because of the bookending rule. Because the bookending rule was removed, the initial resolution step became unnecessary. See https://github.com/json-schema-org/json-schema-spec/issues/1140 for the discussion.

> This PR is labelled "Does this PR introduce a breaking change? No", which seems false now.

Removing the initial resolution step is already a breaking change. That label is wrong regardless of whether we change `$dynamicRef` to not be a URI.